### PR TITLE
[Merged by Bors] - feat(Algebra/Order/SuccPred): typeclass for `succ x = x + 1`

### PR DESCRIPTION
**Pull request by vihdzp**
We create the `SuccAddOrder` typeclass for `SuccOrder`s satisfying `succ x = x + 1`. The instances of this are `ℕ`, `ℤ`, `ℕ∞`, and `Ordinal` (as well as the type alias `NatOrdinal`). 

On such types, `x + 1` should be the preferred form to write `succ x`. This typeclass avoids the need of having to rewrite `x + 1` as `succ x` and back just to apply the common lemmas on `SuccOrder`.

Moreover, there's been talk on [Zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/weaker.20SuccOrder.2FPredOrder/near/467548394) about generalizing `Order.succ` and `Order.pred` to arbitrary preorders. This would make it so that we could no longer set the def-eq `succ x = x + 1`, but having this typeclass means that's not really an issue.

We dually define the `PredSubOrder` typeclass for `pred x = x - 1`. This one is admittedly quite more niche, with the only instances being `ℕ` and `ℤ`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Order/SuccPred/Arithmetic.lean
Comment (vihdzp) [inline Mathlib/Order/SuccPred/Arithmetic.lean:68]: There's some theorems like these that really only apply to `ℤ`, and which won't be very useful anyways since the simp-normal form of `Order.pred` for ℤ is `Int.pred`. Do I just go ahead and remove them? Or do we want them for symmetry?

File: Mathlib/Order/SuccPred/Arithmetic.lean
Comment (YaelDillies) [inline Mathlib/Order/SuccPred/Arithmetic.lean:97]: I see this new API as a way to have an algebraic successor/predecessor API. I see no point stating *any* algebraic lemma about `succ`/`pred`! You should instead be stating order theoretic lemmas about `+ 1`/`- 1`.

File: Mathlib/Order/SuccPred/Arithmetic.lean
Comment (vihdzp) [reply Mathlib/Order/SuccPred/Arithmetic.lean:4]: Sure, I don't mind.

File: Mathlib/Order/SuccPred/Arithmetic.lean
Comment (vihdzp) [reply Mathlib/Order/SuccPred/Arithmetic.lean]: Most of these are generalizations of lemmas that already existed for other types. Is the idea that `x + 1` should be the simp-normal form for `succ x` moving forward?

Comment (vihdzp) [reply Mathlib/Order/SuccPred/Arithmetic.lean]: In fact, maybe it might be a good idea to do the ordinal refactor first. That way we can register `succ_eq_add_one` as a `simp` lemma without it causing conflicts.

Comment (vihdzp) [reply Mathlib/Order/SuccPred/Arithmetic.lean]: Thought about it. It's definitely better to do this refactor first, since we can then use the new lemmas to rewrite the proofs during the ordinal refactor.

Comment (vihdzp) [reply Mathlib/Order/SuccPred/Arithmetic.lean]: I kept `succ_zero` since it was useful in proving other lemmas like `1 ≤ x ↔ 0 < x`.

Comment (YaelDillies) [reply Mathlib/Order/SuccPred/Arithmetic.lean]: > Is the idea that `x + 1` should be the simp-normal form for `succ x` moving forward?

Yes, at least that's the point of this typeclass

File: Mathlib/Data/Int/SuccPred.lean
Comment (YaelDillies) [inline Mathlib/Data/Int/SuccPred.lean:47]: Should this one be deprecated as well?

File: Mathlib/Data/Int/SuccPred.lean
Comment (YaelDillies) [inline Mathlib/Data/Int/SuccPred.lean:50]: And this one too

File: Mathlib/Algebra/Order/SuccPred.lean
Comment (YaelDillies) [inline Mathlib/Algebra/Order/SuccPred.lean:101]: I still believe you can get rid of this lemma...

File: Mathlib/Algebra/Order/SuccPred.lean
Comment (YaelDillies) [inline Mathlib/Algebra/Order/SuccPred.lean:133]: ... eg like this
```suggestion
  simp [← succ_le_iff_of_not_isMax not_isMax_zero, succ_eq_add_one]
```

File: Mathlib/Algebra/Order/SuccPred.lean
Comment (YaelDillies) [inline Mathlib/Algebra/Order/SuccPred.lean:46]: I would make this lemma simp so that people know not to use `succ` when `SuccAddOrder` is around
```suggestion
@[simp] theorem succ_eq_add_one (x : α) : succ x = x + 1 :=
  SuccAddOrder.succ_eq_add_one x
```

File: Mathlib/Data/Int/SuccPred.lean
Comment (vihdzp) [reply Mathlib/Data/Int/SuccPred.lean:47]: This theorem can be used to immediately prove both `Nat.succ_iterate` and `Int.succ_iterate`, which I think is pretty cool. It could potentially be used for stuff on archimedean orders, and it follows our design of using arithmetic instead of `succ`.

File: Mathlib/Data/Int/SuccPred.lean
Comment (vihdzp) [reply Mathlib/Data/Int/SuccPred.lean:50]: I think this one should be kept iff we keep `succ_iterate`.

File: Mathlib/Algebra/Order/SuccPred.lean
Comment (vihdzp) [reply Mathlib/Algebra/Order/SuccPred.lean:46]: Yeah, this would cause a bunch of issues with `Ordinal` as the simp normal form is currently set the other way around. That's a refactor I'll do after this PR gets merged.

File: Mathlib/Algebra/Order/SuccPred.lean
Comment (vihdzp) [reply Mathlib/Algebra/Order/SuccPred.lean:101]: Why do we want to get rid of it? It fits the design of "use arithmetic instead of `succ`" and it's a convenient shortcut for `succ_eq_add_one + zero_add`.

File: Mathlib/Data/Int/SuccPred.lean
Comment (vihdzp) [reply Mathlib/Data/Int/SuccPred.lean:47]: Ah, misread. Well, `Nat.succ_iterate` isn't actually syntactically equal to `Order.succ_iterate`, since one uses `Order.succ` and the other uses `Nat.succ`.

File: Mathlib/Algebra/Order/SuccPred.lean
Comment (YaelDillies) [reply Mathlib/Algebra/Order/SuccPred.lean:101]: > It fits the design of "use arithmetic instead of `succ`" and it's a convenient shortcut for `succ_eq_add_one + zero_add`.

Well clearly it doesn't since it uses `succ`! Once you de`succ`-ify it, you get `(0 : α) + 1 = 1`, which is a purely algebraic lemma

File: Mathlib/Data/Int/SuccPred.lean
Comment (vihdzp) [reply Mathlib/Data/Int/SuccPred.lean:47]: I didn't know that was the plan! It makes a lot of sense though, it's not even a constructor like `Nat.succ` is. Will deprecate these both.

`Order.succ_iterate` and `Order.pred_iterate` can stay though, right?

Comment (YaelDillies) [reply Mathlib/Data/Int/SuccPred.lean:47]: Yes, the plan is to make them replace the above lemma 😉

File: Mathlib/Algebra/Order/SuccPred.lean
Comment (vihdzp) [reply Mathlib/Algebra/Order/SuccPred.lean]: Alright, I'll remove it.

File: Mathlib/SetTheory/Ordinal/Basic.lean
Comment (YaelDillies) [inline Mathlib/SetTheory/Ordinal/Basic.lean:916]:
```suggestion
instance instSuccAddOrder : SuccAddOrder Ordinal := ⟨fun _ => rfl⟩
```

**Review by YaelDillies — APPROVED** [review]
maintainer merge

File: Mathlib/SetTheory/Ordinal/Basic.lean
Comment (vihdzp) [reply Mathlib/SetTheory/Ordinal/Basic.lean:916]: Should the `succOrder` instance just above get renamed too?

Comment (urkud) [reply Mathlib/SetTheory/Ordinal/Basic.lean:916]: Yes, please.

Comment (urkud) [reply Mathlib/SetTheory/Ordinal/Basic.lean:916]: Otherwise LGTM. Thanks!
bors d+

**Pull request closed without merging**
